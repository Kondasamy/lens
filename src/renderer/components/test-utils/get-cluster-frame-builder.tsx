/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainer } from "@ogre-tools/injectable";
import type { LensRendererExtension } from "../../../extensions/lens-renderer-extension";
import rendererExtensionsInjectable from "../../../extensions/renderer-extensions.injectable";
import currentlyInClusterFrameInjectable from "../../routes/currently-in-cluster-frame.injectable";
import { extensionRegistratorInjectionToken } from "../../../extensions/extension-loader/extension-registrator-injection-token";
import { computed } from "mobx";
import { renderFor } from "./renderFor";
import observableHistoryInjectable from "../../navigation/observable-history.injectable";
import currentRouteComponentInjectable from "../../routes/current-route-component.injectable";
import { Sidebar } from "../layout/sidebar";
import React from "react";
import { Router } from "react-router";
import { Observer } from "mobx-react";
import subscribeStoresInjectable from "../../kube-watch-api/subscribe-stores.injectable";
import allowedResourcesInjectable from "../../../common/cluster-store/allowed-resources.injectable";
import type { RenderResult } from "@testing-library/react";

interface Options {
  di: DiContainer;
  extensions: LensRendererExtension[];
}

type Callback = () => void;

export interface ClusterFrameBuilder {
  beforeSetups: (callback: Callback) => ClusterFrameBuilder;
  beforeRender: (callback: Callback) => ClusterFrameBuilder;
  render: () => Promise<RenderResult>;
}

export const getClusterFrameBuilder = ({ di, extensions = [] }: Options) => {
  const beforeSetupsCallbacks: Callback[] = [];
  const beforeRenderCallbacks: Callback[] = [];

  di.override(subscribeStoresInjectable, () => () => () => {});

  di.override(rendererExtensionsInjectable, () => computed(() => extensions));

  di.override(currentlyInClusterFrameInjectable, () => true);

  di.override(allowedResourcesInjectable, () =>
    computed(() => new Set<string>()),
  );

  const builder: ClusterFrameBuilder = {
    beforeSetups(callback: () => void) {
      beforeSetupsCallbacks.push(callback);

      return builder;
    },

    beforeRender(callback: () => void) {
      beforeRenderCallbacks.push(callback);

      return builder;
    },

    async render() {
      beforeSetupsCallbacks.forEach(callback => callback());

      await di.runSetups();

      const extensionRegistrators = di.injectMany(
        extensionRegistratorInjectionToken,
      );

      await Promise.all(
        extensions.flatMap((extension) =>
          extensionRegistrators.map((registrator) => registrator(extension, 1)),
        ),
      );

      const render = renderFor(di);

      const history = di.inject(observableHistoryInjectable);
      const currentRouteComponent = di.inject(currentRouteComponentInjectable);

      beforeRenderCallbacks.forEach(callback => callback());

      return render(
        <Router history={history}>
          <Sidebar />

          <Observer>
            {() => {
              const Component = currentRouteComponent.get();

              if (!Component) {
                return null;
              }

              return <Component />;
            }}
          </Observer>
        </Router>,
      );
    },
  };

  return builder;
};
