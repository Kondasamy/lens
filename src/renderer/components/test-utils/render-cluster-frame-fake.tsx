/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainer } from "@ogre-tools/injectable";
import type { LensRendererExtension } from "../../../extensions/lens-renderer-extension";
import rendererExtensionsInjectable from "../../../extensions/renderer-extensions.injectable";
import currentlyInClusterFrameInjectable from "../../routes/currently-in-cluster-frame.injectable";
import type { Cluster } from "../../../common/cluster/cluster";
import hostedClusterInjectable from "../../../common/cluster-store/hosted-cluster.injectable";
import { extensionRegistratorInjectionToken } from "../../../extensions/extension-loader/extension-registrator-injection-token";
import { computed } from "mobx";
import { renderFor } from "./renderFor";
import observableHistoryInjectable from "../../navigation/observable-history.injectable";
import currentRouteComponentInjectable from "../../routes/current-route-component.injectable";
import { Sidebar } from "../layout/sidebar";
import React from "react";
import { Router } from "react-router";
import { Observer } from "mobx-react";
import subscribeStoresInjectable from "../../kube-watch-api/subscribe-stores.injectable";

interface Options {
  di: DiContainer;
  extensions: LensRendererExtension[];
}

export const renderClusterFrameFake = async ({ di, extensions = [] }: Options) => {
  di.override(subscribeStoresInjectable, () => () => () => {});

  di.override(rendererExtensionsInjectable, () => computed(() => extensions));

  di.override(currentlyInClusterFrameInjectable, () => true);

  const clusterStub = { allowedResources: [] } as Cluster;

  di.override(hostedClusterInjectable, () => clusterStub);

  await di.runSetups();

  const extensionRegistrators = di.injectMany(
    extensionRegistratorInjectionToken,
  );

  await Promise.all(
    extensions.flatMap((extension) =>
      extensionRegistrators.map((registrator) => registrator(extension, 1)),
    ),
  );

  const render = renderFor(di);

  const history = di.inject(observableHistoryInjectable);
  const currentRouteComponent = di.inject(currentRouteComponentInjectable);

  return () => render(
    <Router history={history}>
      <Sidebar />

      <Observer>
        {() => {
          const Component = currentRouteComponent.get();

          if (!Component) {
            return null;
          }

          return <Component />;
        }}
      </Observer>
    </Router>,
  );
};
