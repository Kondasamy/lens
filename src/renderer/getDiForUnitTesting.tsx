/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import glob from "glob";
import {
  join,
  memoize,
  split,
  noop,
  slice,
  filter,
  last,
  map,
  tap,
} from "lodash/fp";
import { createContainer } from "@ogre-tools/injectable";
import { setLegacyGlobalDiForExtensionApi } from "../extensions/as-legacy-globals-for-extension-api/legacy-global-di-for-extension-api";
import getValueFromRegisteredChannelInjectable from "./app-paths/get-value-from-registered-channel/get-value-from-registered-channel.injectable";
import loggerInjectable from "../common/logger.injectable";
import fsInjectable from "../common/fs/fs.injectable";
import type fse from "fs-extra";
import { pipeline } from "@ogre-tools/fp";

export const getDiForUnitTesting = (
  { doGeneralOverrides } = { doGeneralOverrides: false },
) => {
  const di = createContainer();

  setLegacyGlobalDiForExtensionApi(di);

  for (const filePath of getInjectableFilePaths()) {
    const injectableInstance = require(filePath).default;

    di.register({
      ...injectableInstance,
      aliases: [injectableInstance, ...(injectableInstance.aliases || [])],
    });
  }

  di.preventSideEffects();

  if (doGeneralOverrides) {
    di.override(getValueFromRegisteredChannelInjectable, () => () => undefined);

    di.override(fsInjectable, () => getFsFake());

    di.override(loggerInjectable, () => ({
      warn: noop,
      debug: noop,
      log: noop,
      error: (...args: any) => console.error(...args),
      info: noop,
    }));
  }

  return di;
};

const getInjectableFilePaths = memoize(() => [
  ...glob.sync("./**/*.injectable.{ts,tsx}", { cwd: __dirname }),
  ...glob.sync("../common/**/*.injectable.{ts,tsx}", { cwd: __dirname }),
  ...glob.sync("../extensions/**/*.injectable.{ts,tsx}", { cwd: __dirname }),
]);

const getFsFake = () => {
  const state = new Map();

  const readFile = (filePath: string) => {
    const fileContent = state.get(filePath);

    if (!fileContent) {
      const existingFilePaths = [...state.keys()].join('", "');

      throw new Error(
        `Tried to access file ${filePath} which does not exist. Existing file paths are: "${existingFilePaths}"`,
      );
    }

    return Promise.resolve(Buffer.from(fileContent));
  };

  return {
    ensureDir: () => Promise.resolve(),

    writeJson: (filePath: string, contents: object) => {
      state.set(filePath, JSON.stringify(contents));

      return Promise.resolve();
    },

    readJson: async (filePath: string) => {
      const fileContent = await readFile(filePath);

      return JSON.parse(fileContent.toString());
    },

    readFile,

    readdir: (directoryPath: string) => {
      const fileNames = pipeline(
        [...state.keys()],

        filter((x) =>
          pipeline(
            x,
            split("/"),
            slice(0, -1),
            join("/"),
          ),
        ),

        tap(console.log),

        map((x) => pipeline(x, split("/"), last)),
      );

      return Promise.resolve(fileNames);
    },
  } as unknown as typeof fse;
};
