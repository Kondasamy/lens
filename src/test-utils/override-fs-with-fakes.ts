/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainer } from "@ogre-tools/injectable";
import readFileInjectable from "../common/fs/read-file.injectable";
import writeJsonFileInjectable from "../common/fs/write-json-file.injectable";
import readJsonFileInjectable from "../common/fs/read-json-file.injectable";
import pathExistsInjectable from "../common/fs/path-exists.injectable";

export const overrideFsWithFakes = (di: DiContainer) => {
  const state = new Map();

  const readFile = readFileFor(state);

  const overrides = [
    {
      injectableToOverride: readFileInjectable,

      instantiateFake: () => readFile,
    },

    {
      injectableToOverride: writeJsonFileInjectable,

      instantiateFake: () => (filePath: string, contents: object) => {
        state.set(filePath, JSON.stringify(contents));

        return Promise.resolve();
      },
    },

    {
      injectableToOverride: readJsonFileInjectable,

      instantiateFake: () => async (filePath: string) => {
        const fileContent = await readFile(filePath);

        return JSON.parse(fileContent.toString());
      },
    },

    {
      injectableToOverride: pathExistsInjectable,

      instantiateFake: () => async (filePath: string) =>
        Promise.resolve(state.has(filePath)),
    },
  ];

  overrides.forEach(({ injectableToOverride, instantiateFake }) => {
    di.override(injectableToOverride, instantiateFake);
  });
};

const readFileFor = (state: Map<string, Buffer>) => (filePath: string) => {
  const fileContent = state.get(filePath);

  if (!fileContent) {
    const existingFilePaths = [...state.keys()].join('", "');

    throw new Error(
      `Tried to access file ${filePath} which does not exist. Existing file paths are: "${existingFilePaths}"`,
    );
  }

  return Promise.resolve(Buffer.from(fileContent));
};
