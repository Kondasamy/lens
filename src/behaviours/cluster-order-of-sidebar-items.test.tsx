/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { DiContainer, getInjectable } from "@ogre-tools/injectable";
import { getDiForUnitTesting } from "../renderer/getDiForUnitTesting";
import { fireEvent, RenderResult } from "@testing-library/react";
import { renderClusterFrameFake } from "../renderer/components/test-utils/render-cluster-frame-fake";
import {
  SidebarItemRegistration,
  sidebarItemsInjectionToken,
} from "../renderer/components/layout/sidebar-items.injectable";
import { computed } from "mobx";
import { get, includes, noop } from "lodash/fp";
import directoryForLensLocalStorageInjectable from "../common/directory-for-lens-local-storage/directory-for-lens-local-storage.injectable";

describe("cluster order of sidebar items", () => {
  let di: DiContainer;
  let rendered: RenderResult;

  beforeEach(() => {
    di = getDiForUnitTesting({ doGeneralOverrides: true });

    di.override(directoryForLensLocalStorageInjectable, () => "/irrelevant");
  });

  describe("given core registrations", () => {
    let render: () => RenderResult;

    beforeEach(async () => {
      const sidebarItemsInjectable = getInjectable({
        id: "some-sidebar-item-injectable",

        instantiate: () => {
          return computed((): SidebarItemRegistration[] => {
            return [
              {
                id: "some-parent-id",
                parentId: null,
                title: "A Some parent",
                onClick: noop,
                priority: 42,
              },
              {
                id: "some-other-parent-id",
                parentId: null,
                title: "C Some other parent",
                onClick: noop,
                priority: 84,
              },
              {
                id: "some-another-parent-id",
                parentId: null,
                title: "b Some another parent",
                onClick: noop,
                priority: 84,
              },
              {
                id: "some-child-id",
                parentId: "some-parent-id",
                title: "A Some child",
                onClick: noop,
                priority: 126,
              },
              {
                id: "some-other-child-id",
                parentId: "some-parent-id",
                title: "C Some other child",
                onClick: noop,
                priority: 126,
              },
              {
                id: "some-another-child-id",
                parentId: "some-parent-id",
                title: "b Some another child",
                onClick: noop,
                priority: 126,
              },
            ];
          });
        },

        injectionToken: sidebarItemsInjectionToken,
      });

      di.register(sidebarItemsInjectable);

      render = await renderClusterFrameFake({
        di,
        extensions: [],
      });
    });

    describe("when rendered", () => {
      beforeEach(() => {
        rendered = render();
      });

      it("renders", () => {
        expect(rendered.container).toMatchSnapshot();
      });

      it("has parent items in first priority order and then alphabetical", () => {
        const actual = rendered
          .queryAllByTestId("sidebar-item")

          .filter((element) =>
            includes(element.dataset.idTest)([
              "some-parent-id",
              "some-other-parent-id",
              "some-another-parent-id",
            ]),
          )

          .map(get("dataset.titleTest"));

        expect(actual).toEqual([
          "A Some parent",
          "b Some another parent",
          "C Some other parent",
        ]);
      });

      describe("when parent is expanded", () => {
        beforeEach(() => {
          const parentLink = rendered.getByTestId(
            "sidebar-item-link-for-some-parent-id",
          );

          fireEvent.click(parentLink);
        });

        it("renders", () => {
          expect(rendered.container).toMatchSnapshot();
        });

        it("has child items in first priority order and then alphabetical", () => {
          const actual = rendered
            .queryAllByTestId("sidebar-item")
            .filter((element) => element.dataset.parentIdTest === "some-parent-id")
            .map(get("dataset.titleTest"));

          expect(actual).toEqual([
            "A Some child",
            "b Some another child",
            "C Some other child",
          ]);
        });
      });
    });
  });
});
