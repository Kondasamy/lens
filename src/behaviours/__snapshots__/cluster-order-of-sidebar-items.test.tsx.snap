// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cluster order of sidebar items given core registrations when rendered renders 1`] = `
<div>
  <div
    class="flex flex-col"
    data-testid="cluster-sidebar"
  >
    <div>
      <div
        class=""
        style="width: 40px; height: 40px;"
      >
        ??
      </div>
      <div />
    </div>
    <div>
      <div
        class="SidebarItem"
        data-id-test="cluster-overview"
        data-is-active-test="false"
        data-test-id="cluster-overview"
        data-testid="sidebar-item"
        data-title-test="Cluster"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-cluster-overview"
          href="/"
        >
          <i
            class="Icon svg focusable"
          >
            <span
              class="icon"
            >
              [object Object]
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Cluster
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="nodes"
        data-is-active-test="false"
        data-test-id="nodes"
        data-testid="sidebar-item"
        data-title-test="Nodes"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-nodes"
          href="/"
        >
          <i
            class="Icon svg focusable"
          >
            <span
              class="icon"
            >
              [object Object]
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Nodes
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="workloads"
        data-is-active-test="false"
        data-test-id="workloads"
        data-testid="sidebar-item"
        data-title-test="Workloads"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-workloads"
          href="/"
        >
          <i
            class="Icon svg focusable"
          >
            <span
              class="icon"
            >
              [object Object]
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Workloads
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="config"
        data-is-active-test="false"
        data-test-id="config"
        data-testid="sidebar-item"
        data-title-test="Config"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-config"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="list"
            >
              list
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Config
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="some-parent-id"
        data-is-active-test="false"
        data-test-id="some-parent-id"
        data-testid="sidebar-item"
        data-title-test="A Some parent"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-some-parent-id"
          href="/"
        >
          <span
            class="link-text box grow"
          >
            A Some parent
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="network"
        data-is-active-test="false"
        data-test-id="network"
        data-testid="sidebar-item"
        data-title-test="Network"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-network"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="device_hub"
            >
              device_hub
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Network
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="storage"
        data-is-active-test="false"
        data-test-id="storage"
        data-testid="sidebar-item"
        data-title-test="Storage"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-storage"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="storage"
            >
              storage
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Storage
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="namespaces"
        data-is-active-test="false"
        data-test-id="namespaces"
        data-testid="sidebar-item"
        data-title-test="Namespaces"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-namespaces"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="layers"
            >
              layers
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Namespaces
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="events"
        data-is-active-test="false"
        data-test-id="events"
        data-testid="sidebar-item"
        data-title-test="Events"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-events"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="access_time"
            >
              access_time
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Events
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="some-another-parent-id"
        data-is-active-test="false"
        data-test-id="some-another-parent-id"
        data-testid="sidebar-item"
        data-title-test="b Some another parent"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-some-another-parent-id"
          href="/"
        >
          <span
            class="link-text box grow"
          >
            b Some another parent
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="some-other-parent-id"
        data-is-active-test="false"
        data-test-id="some-other-parent-id"
        data-testid="sidebar-item"
        data-title-test="C Some other parent"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-some-other-parent-id"
          href="/"
        >
          <span
            class="link-text box grow"
          >
            C Some other parent
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="helm"
        data-is-active-test="false"
        data-test-id="helm"
        data-testid="sidebar-item"
        data-title-test="Helm"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-helm"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="apps"
            >
              apps
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Helm
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="user-management"
        data-is-active-test="false"
        data-test-id="user-management"
        data-testid="sidebar-item"
        data-title-test="Access Control"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-user-management"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="security"
            >
              security
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Access Control
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="custom-resources"
        data-is-active-test="false"
        data-test-id="custom-resources"
        data-testid="sidebar-item"
        data-title-test="Custom Resources"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-custom-resources"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="extension"
            >
              extension
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Custom Resources
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`cluster order of sidebar items given core registrations when rendered when parent is expanded renders 1`] = `
<div>
  <div
    class="flex flex-col"
    data-testid="cluster-sidebar"
  >
    <div>
      <div
        class=""
        style="width: 40px; height: 40px;"
      >
        ??
      </div>
      <div />
    </div>
    <div>
      <div
        class="SidebarItem"
        data-id-test="cluster-overview"
        data-is-active-test="false"
        data-test-id="cluster-overview"
        data-testid="sidebar-item"
        data-title-test="Cluster"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-cluster-overview"
          href="/"
        >
          <i
            class="Icon svg focusable"
          >
            <span
              class="icon"
            >
              [object Object]
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Cluster
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="nodes"
        data-is-active-test="false"
        data-test-id="nodes"
        data-testid="sidebar-item"
        data-title-test="Nodes"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-nodes"
          href="/"
        >
          <i
            class="Icon svg focusable"
          >
            <span
              class="icon"
            >
              [object Object]
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Nodes
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="workloads"
        data-is-active-test="false"
        data-test-id="workloads"
        data-testid="sidebar-item"
        data-title-test="Workloads"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-workloads"
          href="/"
        >
          <i
            class="Icon svg focusable"
          >
            <span
              class="icon"
            >
              [object Object]
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Workloads
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="config"
        data-is-active-test="false"
        data-test-id="config"
        data-testid="sidebar-item"
        data-title-test="Config"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-config"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="list"
            >
              list
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Config
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="some-parent-id"
        data-is-active-test="false"
        data-test-id="some-parent-id"
        data-testid="sidebar-item"
        data-title-test="A Some parent"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-some-parent-id"
          href="/"
        >
          <span
            class="link-text box grow"
          >
            A Some parent
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_up"
            >
              keyboard_arrow_up
            </span>
          </i>
        </a>
        <ul
          class="sub-menu"
        >
          <div
            class="SidebarItem"
            data-id-test="some-child-id"
            data-is-active-test="false"
            data-parent-id-test="some-parent-id"
            data-test-id="some-child-id"
            data-testid="sidebar-item"
            data-title-test="A Some child"
          >
            <a
              class="nav-item flex gaps align-center"
              data-testid="sidebar-item-link-for-some-child-id"
              href="/"
            >
              <span
                class="link-text box grow"
              >
                A Some child
              </span>
            </a>
          </div>
          <div
            class="SidebarItem"
            data-id-test="some-another-child-id"
            data-is-active-test="false"
            data-parent-id-test="some-parent-id"
            data-test-id="some-another-child-id"
            data-testid="sidebar-item"
            data-title-test="b Some another child"
          >
            <a
              class="nav-item flex gaps align-center"
              data-testid="sidebar-item-link-for-some-another-child-id"
              href="/"
            >
              <span
                class="link-text box grow"
              >
                b Some another child
              </span>
            </a>
          </div>
          <div
            class="SidebarItem"
            data-id-test="some-other-child-id"
            data-is-active-test="false"
            data-parent-id-test="some-parent-id"
            data-test-id="some-other-child-id"
            data-testid="sidebar-item"
            data-title-test="C Some other child"
          >
            <a
              class="nav-item flex gaps align-center"
              data-testid="sidebar-item-link-for-some-other-child-id"
              href="/"
            >
              <span
                class="link-text box grow"
              >
                C Some other child
              </span>
            </a>
          </div>
        </ul>
      </div>
      <div
        class="SidebarItem"
        data-id-test="network"
        data-is-active-test="false"
        data-test-id="network"
        data-testid="sidebar-item"
        data-title-test="Network"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-network"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="device_hub"
            >
              device_hub
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Network
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="storage"
        data-is-active-test="false"
        data-test-id="storage"
        data-testid="sidebar-item"
        data-title-test="Storage"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-storage"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="storage"
            >
              storage
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Storage
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="namespaces"
        data-is-active-test="false"
        data-test-id="namespaces"
        data-testid="sidebar-item"
        data-title-test="Namespaces"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-namespaces"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="layers"
            >
              layers
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Namespaces
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="events"
        data-is-active-test="false"
        data-test-id="events"
        data-testid="sidebar-item"
        data-title-test="Events"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-events"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="access_time"
            >
              access_time
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Events
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="some-another-parent-id"
        data-is-active-test="false"
        data-test-id="some-another-parent-id"
        data-testid="sidebar-item"
        data-title-test="b Some another parent"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-some-another-parent-id"
          href="/"
        >
          <span
            class="link-text box grow"
          >
            b Some another parent
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="some-other-parent-id"
        data-is-active-test="false"
        data-test-id="some-other-parent-id"
        data-testid="sidebar-item"
        data-title-test="C Some other parent"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-some-other-parent-id"
          href="/"
        >
          <span
            class="link-text box grow"
          >
            C Some other parent
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="helm"
        data-is-active-test="false"
        data-test-id="helm"
        data-testid="sidebar-item"
        data-title-test="Helm"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-helm"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="apps"
            >
              apps
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Helm
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="user-management"
        data-is-active-test="false"
        data-test-id="user-management"
        data-testid="sidebar-item"
        data-title-test="Access Control"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-user-management"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="security"
            >
              security
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Access Control
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="custom-resources"
        data-is-active-test="false"
        data-test-id="custom-resources"
        data-testid="sidebar-item"
        data-title-test="Custom Resources"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-custom-resources"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="extension"
            >
              extension
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Custom Resources
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
    </div>
  </div>
</div>
`;
